From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Sun, 14 Jun 2026 18:10:19 +0800
Subject: [PATCH] Arbor Config System


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 3fac32b420ce4ae74166eac43d38c51c1fd513cf..bf068fae416af6e02d0318a43aa240bccb2b8899 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -3032,6 +3032,17 @@ public final class CraftServer implements Server {
         this.console.addPluginAllowingSleep(plugin.getName(), value);
     }
 
+    // Arbor start - Config API
+    @Override
+    public zone.little.arbor.api.config.ArborConfigsInstance getArborConfigs(String name) {
+        return zone.little.arbor.config.ConfigManager.getConfigs(name);
+    }
+
+    public zone.little.arbor.api.config.ArborConfigBuilder getArborConfigBuilder() {
+        return zone.little.arbor.config.ConfigManager.getBuilder();
+    }
+    // Arbor end - Config API
+
     // Folia start - region TPS API
     /**
      * Gets the TPS from the region which owns the specified location, or {@code null} if no region owns
