--- /dev/null
+++ b/src/main/java/me/earthme/luminol/config/RemovedConfig.java
@@ -1,0 +_,20 @@
+package me.earthme.luminol.config;
+
+import java.lang.annotation.*;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Repeatable(RemovedConfig.List.class)
+public @interface RemovedConfig {
+    String name() default "";
+
+    String[] category() default "";
+
+    boolean transfer() default true;
+
+    Class<?> transferLogic() default BasicTransferLogic.class;
+
+    @Retention(RetentionPolicy.RUNTIME)
+    @interface List {
+        RemovedConfig[] value();
+    }
+}
