--- /dev/null
+++ b/src/main/java/me/earthme/luminol/config/TransferringConfig.java
@@ -1,0 +_,22 @@
+package me.earthme.luminol.config;
+
+import java.lang.annotation.Repeatable;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Repeatable(TransferringConfig.List.class)
+public @interface TransferringConfig {
+    String name() default "";
+
+    String[] category() default "";
+
+    boolean transfer() default true;
+
+    Class<? extends DefaultTransferLogic>[] transferLogic() default {DefaultTransferLogic.class};
+
+    @Retention(RetentionPolicy.RUNTIME)
+    @interface List {
+        TransferringConfig[] value();
+    }
+}
