From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Mon, 27 Jan 2025 13:01:59 +0800
Subject: [PATCH] Tick regions api


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index bf2f7c4959f9f2cd43bbd4d4310e7eedf10cf892..c58e67b20acb528ef6321f70837cdbbc6748163d 100644
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -4719,4 +4719,8 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
             }
         }
     }
+
+    // Arbor start - Tick regions api
+    zone.little.arbor.api.ThreadedRegionizer getThreadedRegionizer();
+    // Arbor end - Tick regions api
 }
