From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Fri, 1 May 2026 22:14:43 +0800
Subject: [PATCH] Add config for waypoint restoration


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index db26642e0da92d81a9ae6bc5d6ea4febadab7080..ef0ee453e9e27a9a0e9b493016f437ac0aa03050 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -1138,7 +1138,7 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
     }
 
     private void updateWaypoint() {
-        ServerWaypointManager manager = ((ServerLevel) getHandle().level()).getWaypointManager();
+        me.earthme.luminol.utils.FoliaServerWaypointManager manager = ((ServerLevel) getHandle().level()).getWaypointManager(); // Add config for waypoint restoration
         manager.untrackWaypoint(getHandle());
         manager.trackWaypoint(getHandle());
     }
