From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Fri, 1 May 2026 22:14:43 +0800
Subject: [PATCH] Add config for waypoint restoration


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 03fea774f874b21e87a1281eb07a6b52ed2c31cd..dc985fb08f911e509721169b77cb7494402801d9 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -1138,7 +1138,7 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
     }
 
     private void updateWaypoint() {
-        ServerWaypointManager manager = ((ServerLevel) getHandle().level()).getWaypointManager();
+        me.earthme.luminol.utils.FoliaServerWaypointManager manager = ((ServerLevel) getHandle().level()).getWaypointManager(); // Add config for waypoint restoration
         manager.untrackWaypoint(getHandle());
         manager.trackWaypoint(getHandle());
     }
