From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Fri, 17 Apr 2026 22:45:58 +0800
Subject: [PATCH] Rebrand to Luminol


diff --git a/net/minecraft/server/Main.java b/net/minecraft/server/Main.java
index cf6152c5eaadf2586b00cad19b56401372f8bde0..95361d70182aa5581b19afb6f1adceab56d27531 100644
--- a/net/minecraft/server/Main.java
+++ b/net/minecraft/server/Main.java
@@ -103,7 +103,7 @@ public class Main {
             if (options.has("jfrProfile")) { // CraftBukkit
                 JvmProfiler.INSTANCE.start(Environment.SERVER);
             }
-
+            me.earthme.luminol.config.ConfigManager.initConfigs(); // Luminol - Luminol config
             io.papermc.paper.plugin.PluginInitializerManager.load(options); // Paper
             Bootstrap.bootStrap();
             Bootstrap.validate();
diff --git a/net/minecraft/server/dedicated/DedicatedServer.java b/net/minecraft/server/dedicated/DedicatedServer.java
index eb6f5748bfb1d145a502fb94806f0ab9fe5898aa..413ed813b89ca561e0431bf55f423d244b5ea674 100644
--- a/net/minecraft/server/dedicated/DedicatedServer.java
+++ b/net/minecraft/server/dedicated/DedicatedServer.java
@@ -282,6 +282,7 @@ public class DedicatedServer extends MinecraftServer implements ServerInterface
         this.paperConfigurations.initializeGlobalConfiguration(this.registryAccess());
         this.paperConfigurations.initializeWorldDefaultsConfiguration(this.registryAccess());
         // Paper end - initialize global and world-defaults configuration
+        me.earthme.luminol.config.ConfigManager.loadConfigFiles(); // Luminol - load config file
         this.server.spark.enableEarlyIfRequested(); // Paper - spark
         // Paper start - fix converting txt to json file; convert old users earlier after PlayerList creation but before file load/save
         if (this.convertOldUsers()) {
