From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Fri, 17 Apr 2026 22:46:57 +0800
Subject: [PATCH] Threading utilities


diff --git a/net/minecraft/world/level/Level.java b/net/minecraft/world/level/Level.java
index 28b98b9d3d0273ad419f1b1dc3b29a12a3b87542..e3a4b3afaac404caf82731acbe603f7bbb14d387 100755
--- a/net/minecraft/world/level/Level.java
+++ b/net/minecraft/world/level/Level.java
@@ -142,6 +142,7 @@ public abstract class Level implements LevelAccessor, AutoCloseable, ca.spottedl
     private final RegistryAccess registryAccess;
     private final DamageSources damageSources;
     private final PalettedContainerFactory palettedContainerFactory;
+    public final me.earthme.luminol.utils.thread.RegionizedBlockableEventLoop regionizedBlockableEventLoop = new me.earthme.luminol.utils.thread.RegionizedBlockableEventLoop(this); // Luminol - Threading utilies
     private final java.util.concurrent.atomic.AtomicLong subTickCount = new java.util.concurrent.atomic.AtomicLong(); //private long subTickCount; // Folia - region threading
 
     // CraftBukkit start
