From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Fri, 17 Apr 2026 23:27:09 +0800
Subject: [PATCH] Temporarily fix teleport yam and pitch


diff --git a/net/minecraft/server/commands/TeleportCommand.java b/net/minecraft/server/commands/TeleportCommand.java
index 8d4a17153c17ab3807919a4301e82797c709cdf4..38dc40472057bbc62677724f17bab17ed511dae1 100755
--- a/net/minecraft/server/commands/TeleportCommand.java
+++ b/net/minecraft/server/commands/TeleportCommand.java
@@ -250,8 +250,10 @@ public class TeleportCommand {
             if (true) {
                 ServerLevel worldFinal = level;
                 Vec3 posFinal = new Vec3(x, y, z);
-                Float yawFinal = Float.valueOf(newYRot);
-                Float pitchFinal = Float.valueOf(newXRot);
+                //Float yawFinal = Float.valueOf(newYRot); // Luminol - fix teleport yaw issue
+                //Float pitchFinal = Float.valueOf(newXRot); // Luminol - fix teleport yaw issue
+                Float yawFinal = Float.valueOf(newYRot + victim.getYRot()); // Luminol - fix teleport yaw issue
+                Float pitchFinal = Float.valueOf(newXRot + victim.getXRot()); // Luminol - fix teleport pitch issue
                 victim.getBukkitEntity().taskScheduler.schedule((Entity nmsEntity) -> {
                     nmsEntity.stopRiding();
                     nmsEntity.teleportAsync(
