From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Fri, 31 Jan 2025 20:28:45 +0800
Subject: [PATCH] Add missing teleportation apis for folia


diff --git a/net/minecraft/server/level/ServerPlayer.java b/net/minecraft/server/level/ServerPlayer.java
index 52d70e8ec2ac9b0006483135e2afed174f3939b6..0271277e244bba2bf46154782046c545f766c053 100644
--- a/net/minecraft/server/level/ServerPlayer.java
+++ b/net/minecraft/server/level/ServerPlayer.java
@@ -1640,6 +1640,9 @@ public class ServerPlayer extends Player implements ca.spottedleaf.moonrise.patc
                     if (respawnComplete != null) {
                         respawnComplete.accept(ServerPlayer.this);
                     }
+                    // Luminol - Add missing teleportation apis
+                    new me.earthme.luminol.api.entity.player.PostPlayerRespawnEvent(ServerPlayer.this.getBukkitEntity()).callEvent();
+                    // Luminol end
                 }
             );
         });
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index cc1a9d2292eea91ad7046fa35ebe511515ffb6f3..11d995fc1554afa122475fabc904c88a8ca200dd 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4106,6 +4106,31 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         }
 
         // TODO any events that can modify go HERE
+        // Luminol start - Add missing teleportation apis
+        org.bukkit.Location destinationLoc;
+
+        if (pitch == null) {
+            if (yaw == null) {
+                destinationLoc = io.papermc.paper.util.MCUtil.toLocation(destination, pos, Float.NaN, Float.NaN);
+            } else {
+                destinationLoc = io.papermc.paper.util.MCUtil.toLocation(destination, pos, yaw, Float.NaN);
+            }
+        }else {
+            if (yaw == null) {
+                destinationLoc = io.papermc.paper.util.MCUtil.toLocation(destination, pos, Float.NaN, pitch);
+            }else {
+                destinationLoc = io.papermc.paper.util.MCUtil.toLocation(destination, pos, yaw, pitch);
+            }
+        }
+
+        final me.earthme.luminol.api.entity.EntityTeleportAsyncEvent wrapped = new me.earthme.luminol.api.entity.EntityTeleportAsyncEvent(
+                this.getBukkitEntity(),
+                cause,
+                destinationLoc
+        );
+
+        wrapped.callEvent();
+        // Luminol end
 
         // check for same region
         if (destination == this.level()) {
@@ -4222,7 +4247,18 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
             // we just select the spawn position
             case END: {
                 if (destination.getTypeKey() == net.minecraft.world.level.dimension.LevelStem.END) {
-                    BlockPos targetPos = ServerLevel.END_SPAWN_POINT;
+                    BlockPos targetPos = ServerLevel.END_SPAWN_POINT; // Luminol - Rename
+                    // Luminol start - Add missing teleportation apis
+                    final org.bukkit.Location orginalPortalLocation = io.papermc.paper.util.MCUtil.toLocation(origin, originPortal);
+                    final org.bukkit.Location targetPortalLocation = io.papermc.paper.util.MCUtil.toLocation(destination, targetPos);
+
+                    final me.earthme.luminol.api.portal.PortalLocateEvent portalLocateEvent = new me.earthme.luminol.api.portal.PortalLocateEvent(
+                            orginalPortalLocation,
+                            targetPortalLocation
+                    );
+
+                    portalLocateEvent.callEvent();
+                    // Luminol end
                     // need to load chunks so we can create the platform
                     destination.moonrise$loadChunksAsync(
                         targetPos, 16, // load 16 blocks to be safe from block physics
@@ -4243,7 +4279,18 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                         }
                     );
                 } else {
-                    BlockPos spawnPos = destination.getSharedSpawnPos();
+                    BlockPos spawnPos = destination.getSharedSpawnPos(); // Luminol - Rename
+                    // Luminol start - Add missing teleportation apis
+                    final org.bukkit.Location orginalPortalLocation = io.papermc.paper.util.MCUtil.toLocation(origin, originPortal);
+                    final org.bukkit.Location targetPortalLocation = io.papermc.paper.util.MCUtil.toLocation(destination, spawnPos);
+
+                    final me.earthme.luminol.api.portal.PortalLocateEvent portalLocateEvent = new me.earthme.luminol.api.portal.PortalLocateEvent(
+                            orginalPortalLocation,
+                            targetPortalLocation
+                    );
+
+                    portalLocateEvent.callEvent();
+                    // Luminol end
                     // need to load chunk for heightmap
                     destination.moonrise$loadChunksAsync(
                         spawnPos, 0,
@@ -4294,8 +4341,18 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
 
                 WorldBorder destinationBorder = destination.getWorldBorder();
                 double dimensionScale = net.minecraft.world.level.dimension.DimensionType.getTeleportationScale(origin.dimensionType(), destination.dimensionType());
-                BlockPos targetPos = destination.getWorldBorder().clampToBounds(this.getX() * dimensionScale, this.getY(), this.getZ() * dimensionScale);
+                BlockPos targetPos = destination.getWorldBorder().clampToBounds(this.getX() * dimensionScale, this.getY(), this.getZ() * dimensionScale); // Luminol - Rename
+                // Luminol start - Add missing teleportation apis
+                final org.bukkit.Location orginalPortalLocation = io.papermc.paper.util.MCUtil.toLocation(origin, originPortal);
+                final org.bukkit.Location targetPortalLocation = io.papermc.paper.util.MCUtil.toLocation(destination, targetPos);
+
+                final me.earthme.luminol.api.portal.PortalLocateEvent portalLocateEvent = new me.earthme.luminol.api.portal.PortalLocateEvent(
+                        orginalPortalLocation,
+                        targetPortalLocation
+                );
 
+                portalLocateEvent.callEvent();
+                // Luminol end
                 ca.spottedleaf.concurrentutil.completable.CallbackCompletable<BlockUtil.FoundRectangle> portalFound
                     = new ca.spottedleaf.concurrentutil.completable.CallbackCompletable<>();
 
@@ -4432,6 +4489,15 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
         if (!this.canPortalAsync(destination, takePassengers)) {
             return false;
         }
+        // Luminol start - Add missing teleportation events
+        if (!new me.earthme.luminol.api.entity.PreEntityPortalEvent(
+                this.getBukkitEntity(),
+                io.papermc.paper.util.MCUtil.toLocation(this.level, portalPos),
+                destination.getWorld()
+        ).callEvent()) {
+            return false;
+        }
+        // Luminol end
 
         Vec3 initialPosition = this.position();
         ChunkPos initialPositionChunk = new ChunkPos(
@@ -4499,6 +4565,7 @@ public abstract class Entity implements SyncedDataHolder, Nameable, EntityAccess
                     if (teleportComplete != null) {
                         teleportComplete.accept(teleported);
                     }
+                    // Kaiiju end
                 }
             );
         });
diff --git a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
index 4961bf492e692f9278aadd90ac9d55b2f1c50cba..fc529ece01d8a85e600eb8910662010ab6851d82 100644
--- a/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
+++ b/net/minecraft/world/level/block/entity/TheEndGatewayBlockEntity.java
@@ -191,6 +191,18 @@ public class TheEndGatewayBlockEntity extends TheEndPortalBlockEntity {
             portalTile.trySearchForExit(portalWorld, portalPos);
             return false;
         }
+        // Luminol start - Add missing teleportation apis
+        final org.bukkit.Location orginalPortalLocation = io.papermc.paper.util.MCUtil.toLocation(toTeleport.level(), portalPos);
+        final org.bukkit.Location targetPortalLocation = io.papermc.paper.util.MCUtil.toLocation(portalWorld, teleportPos);
+
+        final me.earthme.luminol.api.portal.PortalLocateEvent portalLocateEvent = new me.earthme.luminol.api.portal.PortalLocateEvent(
+                orginalPortalLocation,
+                targetPortalLocation
+        );
+
+        portalLocateEvent.callEvent();
+        // Luminol end
+
 
         // note: we handle the position from the TeleportTransition
         net.minecraft.world.level.portal.TeleportTransition teleport = net.minecraft.world.level.block.EndGatewayBlock.getTeleportTransition(
