From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Fri, 17 Apr 2026 22:47:34 +0800
Subject: [PATCH] Add config for cpu affinity


diff --git a/io/papermc/paper/threadedregions/TickRegionScheduler.java b/io/papermc/paper/threadedregions/TickRegionScheduler.java
index bea08d4a2e3b228e1fbce1b5e70635416ab972e9..7e9fe197dc0c918f33ac1c2c4dc6bb2b9f22781c 100644
--- a/io/papermc/paper/threadedregions/TickRegionScheduler.java
+++ b/io/papermc/paper/threadedregions/TickRegionScheduler.java
@@ -66,7 +66,26 @@ public final class TickRegionScheduler {
 
             @Override
             public Thread newThread(final Runnable run) {
-                final Thread ret = new TickThreadRunner(this.threadGroup, run, "Folia Region Scheduler Thread #" + this.idGenerator.getAndIncrement());
+                // Luminol start - cpu affinity
+                final Runnable actualRun;
+                if (me.earthme.luminol.config.modules.optimizations.CpuAffinityConfig.cpuAffinityEnabled) {
+                    actualRun = new Runnable() {
+                        private boolean affinitySet = false;
+
+                        @Override
+                        public void run() {
+                            if (!this.affinitySet) {
+                                this.affinitySet = true;
+                                net.openhft.affinity.Affinity.setAffinity(me.earthme.luminol.config.modules.optimizations.CpuAffinityConfig.tickRegionAffinityBitSet);
+                            }
+                            run.run();
+                        }
+                    };
+                } else {
+                    actualRun = run;
+                }
+                // Luminol end - cpu affinity
+                final Thread ret = new TickThreadRunner(this.threadGroup, actualRun, "Folia Region Scheduler Thread #" + this.idGenerator.getAndIncrement());
                 ret.setUncaughtExceptionHandler(TickRegionScheduler.this::uncaughtException);
                 return ret;
             }
