From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Mon, 20 Apr 2026 01:10:30 +0800
Subject: [PATCH] Fix entity portal-teleport speed


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 3f938034888eaf2dea4dce34ea13af44846af5c4..dffe317596a715cc851ea2a623970e2f3ee34714 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1505,7 +1505,27 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         foliaProfiler.startTimer(timerId);
         try {
         // Folia end - profiler
+        // Luminol start - Entity portal-teleport speed fix
         if (isActive) { // Paper - EAR 2
+            if (!(entity instanceof Player) && entity.teleportTickType == 2) { // Luminol - after portal compensate tick
+                entity.tick();
+                entity.tick();
+                entity.teleportTickType = 0;
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
+                    return;
+                }
+                if (entity.handlePortal()) {
+                    return;
+                }
+            } else if (!(entity instanceof Player) && entity.teleportTickType == 1) { // Luminol - portal teleport only
+                entity.teleportTickType++;
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
+                    return;
+                }
+                if (entity.handlePortal()) {
+                    return;
+                }
+            } else {
         entity.tick();
         // Folia start - region threading
         if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
@@ -1523,6 +1543,8 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             worldData.portalRateThrottler.increase(); // Luminol - Portal rate limiter
             return;
         }
+        }
+        // Luminol end - Entity portal-teleport speed fix
         // Folia end - region threading
         } else {entity.inactiveTick();} // Paper - EAR 2
         profiler.pop();
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index adb834d2a69f23583af9c311bc4c80e2fd44944c..7b0e410afe96dafe67241b9482f6946f5016cef1 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -387,6 +387,7 @@ public abstract class Entity
     public long activatedTick = Integer.MIN_VALUE;
     public boolean isTemporarilyActive;
     public long activatedImmunityTick = Integer.MIN_VALUE;
+    public int teleportTickType = 0;// Luminol - Entity portal-teleport speed fix
 
     public void inactiveTick() {
     }
@@ -3549,6 +3550,7 @@ public abstract class Entity
         } else {
             if (this.portalProcess == null || !this.portalProcess.isSamePortal(portal)) {
                 this.portalProcess = new PortalProcessor(portal, pos.immutable());
+                this.teleportTickType = 1; // Luminol - Entity portal-teleport speed fix
             } else if (!this.portalProcess.isInsidePortalThisTick()) {
                 this.portalProcess.updateEntryPosition(pos.immutable());
                 this.portalProcess.setAsInsidePortalThisTick(true);
