From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Wed, 21 May 2025 13:04:00 +0800
Subject: [PATCH] Disable end crystal check


diff --git a/net/minecraft/world/level/dimension/end/EndDragonFight.java b/net/minecraft/world/level/dimension/end/EndDragonFight.java
index 2e156694b337760be986fdf1cbf863b0d896ef2d..b55ac3c0d19cf9117368c17687fa34099bfe3533 100644
--- a/net/minecraft/world/level/dimension/end/EndDragonFight.java
+++ b/net/minecraft/world/level/dimension/end/EndDragonFight.java
@@ -546,6 +546,8 @@ public class EndDragonFight {
 
                 blockPos = this.portalLocation;
             }
+            // Luminol start - Disable end crystal check
+            if (!me.earthme.luminol.config.modules.misc.DisableEndCrystalCheckConfig.disableEndCrystalCheck) {
             // Paper start - Perf: Do crystal-portal proximity check before entity lookup
             if (placedEndCrystalPos != null) {
                 // The end crystal must be 0 or 1 higher than the portal origin
@@ -561,6 +563,7 @@ public class EndDragonFight {
                 }
             }
             // Paper end - Perf: Do crystal-portal proximity check before entity lookup
+            } // Luminol end - Disable end crystal check
 
 
             List<EndCrystal> list = Lists.newArrayList();
