From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Mon, 26 May 2025 23:48:32 +0800
Subject: [PATCH] Use old registering way of ticket of scheduler api of folia


diff --git a/net/minecraft/server/level/TicketType.java b/net/minecraft/server/level/TicketType.java
index bbcb0fdf0cdf6f860897fe39d65019de19b3d758..99a5db75ae388caa804eadfcaed307c8bed419df 100644
--- a/net/minecraft/server/level/TicketType.java
+++ b/net/minecraft/server/level/TicketType.java
@@ -57,7 +57,7 @@ public final class TicketType<T> implements ca.spottedleaf.moonrise.patches.chun
     public static final TicketType<Long> END_GATEWAY_EXIT_SEARCH = ca.spottedleaf.moonrise.patches.chunk_system.ticket.ChunkSystemTicketType.create("folia:end_gateway_exit_search", Long::compareTo);
     public static final TicketType<Long> NETHER_PORTAL_DOUBLE_CHECK = ca.spottedleaf.moonrise.patches.chunk_system.ticket.ChunkSystemTicketType.create("folia:nether_portal_double_check", Long::compareTo);
     public static final TicketType<Long> TELEPORT_HOLD_TICKET = ca.spottedleaf.moonrise.patches.chunk_system.ticket.ChunkSystemTicketType.create("folia:teleport_hold_ticket", Long::compareTo);
-    public static final TicketType REGION_SCHEDULER_API_HOLD = register("folia:region_scheduler_api_hold", 0L, false, TicketType.TicketUse.LOADING_AND_SIMULATION);
+    public static final TicketType REGION_SCHEDULER_API_HOLD = ca.spottedleaf.moonrise.patches.chunk_system.ticket.ChunkSystemTicketType.create("folia:region_scheduler_api_hold", (a, b) -> 0); // Luminol - Use old registering way
     // Folia end - region threading
 
     public static TicketType register(String name, long timeout, boolean persist, TicketType.TicketUse use) {
