From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mrhua269 <mrhua269@gmail.com>
Date: Sat, 3 Jan 2026 10:42:39 +0800
Subject: [PATCH] Sync dragon part when teleportation or firstly created

A simple fix of https://github.com/PaperMC/Folia/issues/371

diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index d03795cd025dd36363002770b7a6a47676ffa6bf..1eee289eb288698a419f6c89e81c0acd9e63e1bd 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -4407,6 +4407,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         Entity copy = this.getType().create(destination, EntitySpawnReason.DIMENSION_TRAVEL);
         copy.restoreFrom(this);
         copy.transform(pos, yaw, pitch, velocity);
+        if (copy instanceof net.minecraft.world.entity.boss.enderdragon.EnderDragon dragon) dragon.syncDragonPartsAfterTeleportTransform(); // Luminol - Sync dragon part when teleportation or firstly created
         if (toSameWorld) copy.notifyLithiumTrackerIfNeeded();// Luminol - region threading for sleeping block entity
         // vanilla code used to call remove _after_ copying, and some stuff is required to be after copy - so add hook here
         // for example, clearing of inventory after switching dimensions
diff --git a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
index 1daf8f52148043339f723aaebdf6135763dd805b..b431d27f84c81ebc77926fbe34beb8870675a652 100644
--- a/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
+++ b/net/minecraft/world/entity/boss/enderdragon/EnderDragon.java
@@ -999,4 +999,21 @@ public class EnderDragon extends Mob implements Enemy {
         return flag ? i : 0;
     }
     // CraftBukkit end
+    
+    // Luminol start - Sync dragon part when teleportation or firstly created
+    @Override
+    public net.minecraft.world.entity.SpawnGroupData finalizeSpawn(net.minecraft.world.level.ServerLevelAccessor level, net.minecraft.world.DifficultyInstance difficulty, net.minecraft.world.entity.EntitySpawnReason spawnReason, net.minecraft.world.entity.SpawnGroupData spawnGroupData) {
+        for (EnderDragonPart part : this.subEntities) {
+            this.tickPart(part, 0.0, 0.0, 0.0); // offset -> 0.0
+        }
+
+        return super.finalizeSpawn(level, difficulty, spawnReason, spawnGroupData);
+    }
+
+    public void syncDragonPartsAfterTeleportTransform() {
+        for (EnderDragonPart part : this.subEntities) {
+            this.tickPart(part, 0.0, 0.0, 0.0); // offset -> 0.0
+        }
+    }
+    // Luminol end
 }
