From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: mrhua269 <mrhua269@gmail.com>
Date: Sun, 15 Feb 2026 22:39:13 +0800
Subject: [PATCH] Recompute path when path finding out of current tick region


diff --git a/net/minecraft/world/entity/ai/navigation/PathNavigation.java b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
index 0a77b906a426732a2f82e8727ef99b0ebf692238..25f63b356975c5205eea707a0d102c2ed4b7c2f4 100644
--- a/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -283,6 +283,19 @@ public abstract class PathNavigation {
 
             if (!this.isDone()) {
                 Vec3 tempMobPos = this.path.getNextEntityPos(this.mob);
+                // Luminol - Recompute path when path finding out of current tick region
+                if (me.earthme.luminol.config.modules.fixes.RecomputePathWhenOutOfRegionConfig.enabled) {
+                    // The target seems to be pointed to an position out of current region, so reset if it is
+                    if (this.targetPos != null && !ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.mob.level(), this.targetPos)) {
+                        this.targetPos = null;
+                    }
+
+                    if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(this.mob.level(), tempMobPos)) {
+                        this.hasDelayedRecomputation = true;
+                        return;
+                    }
+                }
+                // Luminol end
                 this.mob.getMoveControl().setWantedPosition(tempMobPos.x, this.getGroundY(tempMobPos), tempMobPos.z, this.speedModifier);
             }
         }
