From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Mon, 20 Apr 2026 01:10:30 +0800
Subject: [PATCH] Fix entity portal-teleport speed


diff --git a/net/minecraft/server/level/ServerLevel.java b/net/minecraft/server/level/ServerLevel.java
index 35ab432bbd59aba80a6eaca1769e6aa82c8a7e46..a9209f285f6df5ab00899a7ec4895f5b6a1a3c43 100644
--- a/net/minecraft/server/level/ServerLevel.java
+++ b/net/minecraft/server/level/ServerLevel.java
@@ -1492,7 +1492,27 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
         foliaProfiler.startTimer(timerId);
         try {
         // Folia end - profiler
+        // Arbor start - Entity portal-teleport speed fix
         if (isActive) { // Paper - EAR 2
+            if (!(entity instanceof Player) && entity.teleportTickType == 2) { // Arbor - after portal compensate tick
+                entity.tick();
+                entity.tick();
+                entity.teleportTickType = 0;
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
+                    return;
+                }
+                if (entity.handlePortal()) {
+                    return;
+                }
+            } else if (!(entity instanceof Player) && entity.teleportTickType == 1) { // Arbor - portal teleport only
+                entity.teleportTickType++;
+                if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
+                    return;
+                }
+                if (entity.handlePortal()) {
+                    return;
+                }
+            } else {
         entity.tick();
         // Folia start - region threading
         if (!ca.spottedleaf.moonrise.common.util.TickThread.isTickThreadFor(entity)) {
@@ -1503,6 +1523,8 @@ public class ServerLevel extends Level implements WorldGenLevel, ServerEntityGet
             // portalled
             return;
         }
+        }
+        // Arbor end - Entity portal-teleport speed fix
         // Folia end - region threading
         } else {entity.inactiveTick();} // Paper - EAR 2
         profiler.pop();
diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index b86381f0e7f14158b67f102dcbf23d5d3bfe063e..4122823afdd54927aa79bbd597993ae3a54dc5c7 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -388,6 +388,7 @@ public abstract class Entity
     public long activatedTick = Integer.MIN_VALUE;
     public boolean isTemporarilyActive;
     public long activatedImmunityTick = Integer.MIN_VALUE;
+    public int teleportTickType = 0;// Arbor - Entity portal-teleport speed fix
 
     public void inactiveTick() {
     }
@@ -3597,6 +3598,7 @@ public abstract class Entity
         } else {
             if (this.portalProcess == null || !this.portalProcess.isSamePortal(portal)) {
                 this.portalProcess = new PortalProcessor(portal, pos.immutable());
+                this.teleportTickType = 1; // Arbor - Entity portal-teleport speed fix
             } else if (!this.portalProcess.isInsidePortalThisTick()) {
                 this.portalProcess.updateEntryPosition(pos.immutable());
                 this.portalProcess.setAsInsidePortalThisTick(true);
