From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <mrhua269@gmail.com>
Date: Sat, 18 Apr 2026 09:08:52 +0800
Subject: [PATCH] Add config for server mod name


diff --git a/net/minecraft/server/MinecraftServer.java b/net/minecraft/server/MinecraftServer.java
index fc1551291d9a5072df98e93d26e6b183ba035c64..d494bd4f6902a7d52923b1c2c001e89021bb040b 100644
--- a/net/minecraft/server/MinecraftServer.java
+++ b/net/minecraft/server/MinecraftServer.java
@@ -2096,7 +2096,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     }
 
     public String getServerModName() {
-        return io.papermc.paper.ServerBuildInfo.buildInfo().brandName(); // Paper
+        return zone.little.arbor.config.modules.misc.ServerModNameConfig.fakeVanilla ? "vanilla" : zone.little.arbor.config.modules.misc.ServerModNameConfig.serverModName; // Paper // Arbor - Add config for server mod name
     }
 
     public ServerClockManager clockManager() {
