From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: MrHua269 <wangxyper@163.com>
Date: Mon, 27 Jan 2025 13:01:59 +0800
Subject: [PATCH] Tick regions api


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
index e2b65c0d2d6b9fb6fb4003a8e57766d410a95322..21eac354badd6974cd1e51233c1c775992807a38 100755
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -4621,4 +4621,8 @@ public interface World extends RegionAccessor, WorldInfo, PluginMessageRecipient
             }
         }
     }
+
+    // Luminol start - Tick regions api
+    me.earthme.luminol.api.ThreadedRegionizer getThreadedRegionizer();
+    // Luminol end - Tick regions api
 }
