From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Helvetica Volubi <suisuroru@blue-millennium.fun>
Date: Sat, 24 May 2025 01:47:10 +0800
Subject: [PATCH] Leaves Fix SculkCatalyst exp skip


diff --git a/src/main/java/org/bukkit/event/entity/PlayerDeathEvent.java b/src/main/java/org/bukkit/event/entity/PlayerDeathEvent.java
index 6316a2f1d2dc0314397e33e6dbd354fb8bc50541..6e2d1bf0e8eb9fa4917b464324a02b9b5d0dd496 100644
--- a/src/main/java/org/bukkit/event/entity/PlayerDeathEvent.java
+++ b/src/main/java/org/bukkit/event/entity/PlayerDeathEvent.java
@@ -25,6 +25,7 @@ public class PlayerDeathEvent extends EntityDeathEvent {
     private boolean doExpDrop;
     private boolean keepLevel = false;
     private boolean keepInventory = false;
+    private boolean useApiExpDropStatus = false; // Leaves - exp fix
     @Deprecated
     private final List<ItemStack> itemsToKeep = new ArrayList<>();
 
@@ -82,8 +83,15 @@ public class PlayerDeathEvent extends EntityDeathEvent {
         this.showDeathMessages = true;
         this.deathMessage = LegacyComponentSerializer.legacySection().deserializeOrNull(deathMessage);
         this.doExpDrop = doExpDrop;
+        this.useApiExpDropStatus = true; // Leaves - exp fix
     }
 
+    // Leaves start - exp fix
+    public boolean forceUseEventDropStatus() {
+        return this.useApiExpDropStatus;
+    }
+    // Leaves end - exp fix
+
     @NotNull
     @Override
     public Player getEntity() {
